feat: update clones if they're present

mortenpi opened the pull request:
So if you are cloning the remotes to a local directory, it updates the local remotes. Not a 100% sure about this, but it would be good to have something like this.

It does slow things down a bit (not as bad as full clones, but still). So maybe we want to guard this with an environment variable or something?

mortenpi commented on src/MultiDocumenter.jl:
```diff
@@ -188,6 +188,17 @@ function maybe_clone(docs::Vector{MultiDocRef})
+            elseif isempty(current_branch)
+                @error "current_branch was empty"
+            end
+            run(`$(git()) -C $(doc.upstream) reset --hard origin/$(doc.branch)`)
```
I am also not super comfortable with the `reset --hard` here. Especially if `doc.upstream` is pointing into a subdirectory of a real directory. Actually, instead of `-C`, should probably use `--git-dir`..

pfitzseb replied:
Is there ever any harm in hard-resetting a remote branch?

mortenpi replied:
I think if everything is normal, then it doesn't do anything wrong. But the cases I'm concerned about:

1. Something has gone horribly wrong, and `doc.upstream` is not actually a separate Git repo anymore, and then this command nukes changes on your main repository.
2. The user has actually gone into the cloned directories and modified stuff, and then we nuke their updates.

So I am wondering if something like a `git merge --ff-only` would be the better option here. I didn't want to do a simple `git pull`, since that, I believe, would fetch all branches.

pfitzseb replied:
What I meant is that `reset --hard origin/mybranch` can't ever nuke any local changes, because those are in `mybranch` instead.

mortenpi replied:
I might be mistaken, but I believe `reset --hard` will change the commit your _current_ branch points to `origina/mybranch` (any commit not on other branches/repos will become lost-ish), and it will also nuke all uncommitted working tree changes.

pfitzseb replied:
Oh, doh! Yeah, I just misread the command here :P

Ok, so how about we just checkout `remote/mybranch` for the build?

mortenpi commented:
I updates this, making it `git checkout --detach origin/$branch` when it updates the branch, and also added a check to see if the local directory has any changes. And I am also passing `-C` and `--git-dir` to all the `git` commands, so I think they should all run only when the clone is a valid, functioning Git clone, and not accidentally run the commands on some parent directory.

Unrelatedly, I also added `--no-tags` to the original `clone` command. It seems like [`--single-branch` does not actually imply that](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---no-tags).
